src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -0,0 +1,229 @@
+#include "sql_pipeline_statement.hpp"
+
+#include <boost/algorithm/string.hpp>
+
+#include <iomanip>
+#include <utility>
+
+#include "SQLParser.h"
+#include "concurrency/transaction_manager.hpp"
+#include "logical_query_plan/lqp_translator.hpp"
+#include "optimizer/optimizer.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "sql/sql_translator.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+SQLPipelineStatement::SQLPipelineStatement(const std::string& sql, bool use_mvcc)
+    : _sql_string(sql), _use_mvcc(use_mvcc), _auto_commit(_use_mvcc) {}
+
+SQLPipelineStatement::SQLPipelineStatement(const std::string& sql,
+                                           std::shared_ptr<TransactionContext> transaction_context)
+    : _sql_string(sql), _use_mvcc(true), _auto_commit(false), _transaction_context(std::move(transaction_context)) {
+  DebugAssert(_transaction_context != nullptr, "Cannot pass nullptr as explicit transaction context.");
+}
+
+SQLPipelineStatement::SQLPipelineStatement(std::shared_ptr<hsql::SQLParserResult> parsed_sql,
+                                           std::shared_ptr<TransactionContext> transaction_context, bool use_mvcc)
+    : _parsed_sql_statement(std::move(parsed_sql)),
+      _use_mvcc(use_mvcc),
+      _auto_commit(_use_mvcc && transaction_context == nullptr) {
+  Assert(_parsed_sql_statement->size() == 1, "SQLPipelineStatement must hold exactly one SQL statement");
+  // We don't want to create a new context yet, as it should contain all changes of previously created (possibly even in
+  // same query) pipelines up to the point of this pipeline's execution.
+  if (transaction_context != nullptr) {
+    _transaction_context = std::move(transaction_context);
+  }
+}
+
+const std::shared_ptr<hsql::SQLParserResult>& SQLPipelineStatement::get_parsed_sql_statement() {
+  if (_parsed_sql_statement) {
+    // Return cached result
+    return _parsed_sql_statement;
+  }
+
+  DebugAssert(!_sql_string.empty(), "Cannot parse empty SQL string");
+
+  _parsed_sql_statement = std::make_shared<hsql::SQLParserResult>();
+  try {
+    hsql::SQLParser::parse(_sql_string, _parsed_sql_statement.get());
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while parsing SQL query:\n  " + std::string(exception.what()));
+  }
+
+  if (!_parsed_sql_statement->isValid()) {
+    throw std::runtime_error(SQLPipelineStatement::create_parse_error_message(_sql_string, *_parsed_sql_statement));
+  }
+
+  Assert(_parsed_sql_statement->size() == 1, "SQLPipelineStatement must hold exactly one statement");
```

Comment:
It is helpful if we include suggested fixes in the assert message whenever it makes sense. Maybe "Use SQLPipeline when you have multiple statements."